File: 0x10 基本数据结构/0x11 栈/AcWing41.py
class MinStack(object):

    def __init__(self):
        self.stk = []
        self.stk_min = []

    def push(self, val: int) -> None:
        self.stk.append(val)
        if self.stk_min:
            val = min(val, self.stk_min[-1])
        self.stk_min.append(val)

    def pop(self) -> None:
        self.stk.pop()
        self.stk_min.pop()

    def top(self) -> int:
        return self.stk[-1]

    def getMin(self) -> int:
        return self.stk_min[-1]
        

# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: 0x10 基本数据结构/0x11 栈/AcWing41.cpp
class MinStack {
public:
  /** initialize your data structure here. */

  stack<int> stk, stk_min; // stk_min：辅助栈，栈顶始终保存当前数据栈中最小值

  MinStack() {

  }

  void push(int x) {
    stk.push(x);
    if (stk_min.size()) x = min(x, stk_min.top());
    stk_min.push(x);
  }

  void pop() {
    stk.pop();
    stk_min.pop();
  }

  int top() {
    return stk.top();
  }

  int getMin() {
    return stk_min.top();
  }
};